runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/stableinput/BufferingDoFnRunner.java

```diff
@@ -0,0 +1,213 @@
+      String stateName,
+      org.apache.beam.sdk.coders.Coder inputCoder,
+      org.apache.beam.sdk.coders.Coder windowCoder,
+      OperatorStateBackend operatorStateBackend,
+      @Nullable KeyedStateBackend keyedStateBackend)
+      throws Exception {
+
+    this.underlying = underlying;
+    this.notYetAcknowledgedSnapshots =
+        operatorStateBackend.getUnionListState(
+            new ListStateDescriptor<>("notYetAcknowledgedSnapshots", CheckpointElement.class));
+    this.bufferingElementsHandlerFactory =
+        (stateId) -> {
+          ListStateDescriptor<BufferedElement> stateDescriptor =
+              new ListStateDescriptor<>(
+                  stateName + stateId,
+                  new CoderTypeSerializer<>(new BufferedElements.Coder(inputCoder, windowCoder)));
+          if (keyedStateBackend != null) {
+            return KeyedBufferingElementsHandler.create(keyedStateBackend, stateDescriptor);
+          } else {
+            return NonKeyedBufferingElementsHandler.create(
+                operatorStateBackend.getListState(stateDescriptor));
+          }
+        };
+    this.currentStateId = generateNewId();
+    this.currentBufferingElementsHandler = bufferingElementsHandlerFactory.create(currentStateId);
+  }
+
+  @Override
+  public void startBundle() {
+    // Do not start a bundle, start it later when emitting elements
+  }
+
+  @Override
+  public void processElement(WindowedValue<InputT> elem) {
+    currentBufferingElementsHandler.buffer(new BufferedElements.Element(elem));
+  }
+
+  @Override
+  public void onTimer(
+      String timerId, BoundedWindow window, Instant timestamp, TimeDomain timeDomain) {
+    currentBufferingElementsHandler.buffer(
+        new BufferedElements.Timer(timerId, window, timestamp, timeDomain));
+  }
+
+  @Override
+  public void finishBundle() {
+    // Do not finish a bundle, finish it later when emitting elements
+  }
+
+  @Override
+  public DoFn<InputT, OutputT> getFn() {
+    return underlying.getFn();
+  }
+
+  /** Should be called when a checkpoint is created. */
+  public void checkpoint(long checkpointId) throws Exception {
+    // We are about to get checkpointed. The elements buffered thus far
+    // have to be added to the global CheckpointElement state which will
+    // be used to emit elements later when this checkpoint is acknowledged.
+    addToBeAcknowledgedCheckpoint(checkpointId, currentStateId);
+    currentStateId = generateNewId();
+    currentBufferingElementsHandler = bufferingElementsHandlerFactory.create(currentStateId);
+  }
+
+  /** Should be called when a checkpoint is completed. */
+  public void checkpointCompleted(long checkpointId) throws Exception {
+    List<CheckpointElement> toAck = removeToBeAcknowledgedCheckpoints(checkpointId);
+    for (CheckpointElement toBeAcked : toAck) {
+      BufferingElementsHandler bufferingElementsHandler =
+          bufferingElementsHandlerFactory.create(toBeAcked.internalId);
```

Comment:
Maybe `get` instead of `create` to convey that we are accessing something that already exists?